zio-cli/shared/src/main/scala/zio/cli/FigFontParser.scala

```diff
@@ -0,0 +1,336 @@
+  object Fitted   extends LayoutMode
+  object Smushed  extends LayoutMode
+}
+
+object FigFontParser {
+  import ParseLib._
+
+  def parse(lines: Seq[String]): Either[String, FigFont] =
+    LinesParsers
+      .parse(LinesParsers.LinesReader(lines), FlfParsers.font)
+      .fold(e => Left(e()), Right(_))
+}
+
+private[cli] object FlfParsers {
+  import ParseLib.LinesParsers
+  import ParseLib.LinesParsers._
+
+  val font: LinesParsers.Parser[FigFont] = for {
+    header <- header
+    _      <- take(header.commentLines) // skip comments
+    chars  <- charsParser(header)
+    // TODO add EOF handling to report broken files
+  } yield FigFont(
+    header,
+    rightToLeft = header.rightToLeft.contains(1),
+    layout = toLayout(header.oldLayoutMask, header.fullLayoutMask),
+    chars = chars
+  )
+
+  lazy val header: LinesParsers.Parser[FigHeader] = for {
+    line   <- one
+    lr     = ParseLib.CharsParsers.CharsReader(line)
+    header <- asParser(ParseLib.CharsParsers.parse(lr, FlfLineParsers.header))
+  } yield header
+
+  def charsParser(header: FigHeader): LinesParsers.Parser[Map[Char, FigChar]] = {
+    val charTag: LinesParsers.Parser[Char] = fail("TODO") // TODO, fold skip tagged figchars this way for now
+
+    val figChar = for {
+      lines <- take(header.charHeight).map(_.lines)
+      first = lines.head
+      endMarker = first.last
+      width = first.indexOf(endMarker.toInt)
+      charLines = lines.map { l =>
+        val cut = l.take(width)
+        val leftSpaces = cut.takeWhile(_ == ' ').count(_ => true)
+        val rightSpaces = cut.reverse.takeWhile(_ == ' ').count(_ => true)
+        FigCharLine(leftSpaces, rightSpaces, cut.drop(leftSpaces).dropRight(rightSpaces))
+      }
+    } yield FigChar(charLines, width, header.charHeight)
+
+    for {
+      required <- figChar.rep(requiredChars.size).map(requiredChars.zip(_).toMap)
+      tagged   <- (charTag ~ figChar).fold[Map[Char, FigChar]](Map.empty)(_ + _)
+    } yield required ++ tagged
+  }
+
+  private val requiredChars: Seq[Char] = (32 until 127).map(_.toChar) ++ "ÄÖÜßäöü".toCharArray
+
+  private def toLayout(oldLayoutMake: Int, fullLayoutMask: Option[Int]): Layout = ???
+}
+
+private[cli] object FlfLineParsers {
+  import ParseLib.CharsParsers
+  import ParseLib.CharsParsers._
+
+  val header: CharsParsers.Parser[FigHeader] = for {
+    signature      <- take(5).map(_.chars.toString)
+    hardBlank      <- take(1).map(_.head)
+    charHeight     <- one(' ') ~> int
+    baseline       <- one(' ') ~> int
+    maxLength      <- one(' ') ~> int
+    oldLayoutMask  <- one(' ') ~> int
+    commentLines   <- one(' ') ~> int
+    rightToLeft    <- optional(one(' ') ~> int)
+    fullLayoutMask <- optional(one(' ') ~> int)
+  } yield FigHeader(
+    signature,
+    hardBlank,
+    charHeight,
+    baseline,
+    maxLength,
+    commentLines,
+    oldLayoutMask,
+    rightToLeft,
+    fullLayoutMask
+  )
+}
+
+private[cli] object ParseLib {
```

Comment:
Was this code imported or adapted from anywhere, such that we should reproduce copyright notices and ensure license compatibility? Or is it novel?

Reply:
Novel, but i looked at https://github.com/scala/scala-parser-combinators for inspiration. I'd like to revisit it again, as the font file parser turned out simpler than i anticipated 😄